[IAP] Track events

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10023 . Ref: pdfdoF-3d1-p2

## Description

This PR adds track events for In-App Purchases related events. Unit Tests will be added on a different PR.

## Changes
* Added the following `WooAnalyticsStats` and `WooAnalyticsEvents`:
  * `plan_upgrade_purchase_button_tapped`
  * `plan_upgrade_screen_loaded`
  * `plan_upgrade_screen_dismissed`
  * `plan_upgrade_feature_screen_loaded`
  * `plan_upgrade_processing_screen_loaded`
  * `plan_upgrade_processing_screen_dismissed`
  * `plan_upgrade_completed_screen_loaded`
  * `plan_upgrade_purchase_failed`
  * `planUpgradePurchaseButtonTapped(productID: String)`
  * `planUpgradeScreenLoaded(source: Source)`
  * `planUpgradeScreenDismissed(step: Step)`
  * `planUpgradeFeatureScreenLoaded(featureGroup: String)`
  * `planUpgradePurchaseFailed(error: InAppPurchasesError)`

## Testing instructions:
#### Unhappy path:

When running the app and attempting the unhappy path, for example a purchase via the WooCommerce IAP scheme, we'll go through the following flow and track the following events:

1. Tap on "Upgrade Now", see tracked event when the screen effectively loads the plan information: 🔵 Tracked plan_upgrade_screen_loaded

```
2023-06-26 11:04:37.984820+0700 WooCommerce[89036:73989858] 🔵 Tracked plan_upgrade_screen_loaded, properties: [AnyHashable("blog_id"): 220379441, AnyHashable("is_wpcom_store"): true]`
```

2. Dismiss the screen:  🔵 Tracked plan_upgrade_screen_dismissed - Step: plan_details

```
2023-06-26 14:56:26.632233+0700 WooCommerce[6629:3154589] 🔵 Tracked plan_upgrade_screen_dismissed, properties: [AnyHashable("blog_id"): 220579378, AnyHashable("is_wpcom_store"): true, AnyHashable("step"): "plan_details"]
```

3. Tap again and scroll and tap on any of the plan properties: 🔵 Tracked plan_upgrade_feature_screen_loaded, with specific feature_group:

```
2023-06-26 14:43:12.727861+0700 WooCommerce[94405:74084810] 🔵 Tracked plan_upgrade_feature_screen_loaded, properties: [AnyHashable("feature_group"): "General features", AnyHashable("blog_id"): 220579378, AnyHashable("is_wpcom_store"): true]
```

4. Tap "Purchase Debug Essential Monthly": 🔵 Tracked plan_upgrade_purchase_button_tapped, with product_ID = debug.woocommerce.express.essential.monthly

```
2023-06-26 09:47:27.383473+0700 WooCommerce[84628:73924741] 🔵 Tracked plan_upgrade_purchase_button_tapped, properties: [AnyHashable("blog_id"): 220379441, AnyHashable("is_wpcom_store"): true, AnyHashable("product_ID"): "debug.woocommerce.express.essential.monthly"]
```

5. When the processing screen starts loading: 🔵 Tracked plan_upgrade_processing_screen_loaded

```
2023-06-26 14:30:59.365559+0700 WooCommerce[93484:74076645] 🔵 Tracked plan_upgrade_processing_screen_loaded, properties: [AnyHashable("is_wpcom_store"): true, AnyHashable("blog_id"): 220379441]
```

6. When the purchase process fails: 🔵 Tracked plan_upgrade_purchase_failed, with a specific error (or unknown if that's the case)

```
2023-06-26 14:31:03.801571+0700 WooCommerce[93484:74076645] 🔵 Tracked plan_upgrade_purchase_failed, properties: [AnyHashable("is_wpcom_store"): true, AnyHashable("blog_id"): 220379441, AnyHashable("error"): "unknown"]
```

7. When the error screen is dismissed: 🔵 Tracked plan_upgrade_processing_screen_dismissed

```
2023-06-26 14:31:03.821092+0700 WooCommerce[93484:74076646] 🔵 Tracked plan_upgrade_processing_screen_dismissed, properties: [AnyHashable("blog_id"): 220379441, AnyHashable("is_wpcom_store"): true]
```

8. We dismiss the purchase screen: 🔵 Tracked plan_upgrade_screen_dismissed

```
2023-06-26 14:33:00.211026+0700 WooCommerce[93484:74077749] 🔵 Tracked plan_upgrade_screen_dismissed, properties: [AnyHashable("step"): "purchase_upgrade_error", AnyHashable("blog_id"): 220379441, AnyHashable("is_wpcom_store"): true]
```

#### Happy path:

If we attempt the purchase via the physical simulator on the WooCommerce scheme, we will see the additional tracks:

1. When the completed screen is shown: 🔵 Tracked plan_upgrade_completed_screen_loaded

```
2023-06-26 14:59:16.031235+0700 WooCommerce[6629:3157076] 🔵 Tracked plan_upgrade_completed_screen_loaded, properties: [AnyHashable("is_wpcom_store"): true, AnyHashable("blog_id"): 220579378]
```

2. When we tap in "Done": 🔵 Tracked plan_upgrade_screen_dismissed - step: Completed

```
2023-06-26 14:59:36.326650+0700 WooCommerce[6629:3157220] 🔵 Tracked plan_upgrade_screen_dismissed, properties: [AnyHashable("blog_id"): 220579378, AnyHashable("is_wpcom_store"): true, AnyHashable("step"): "completed"]
```

iamgabrielma committed:
Add IAP extension to track events

iamgabrielma committed:
Add track case for dismissal

iamgabrielma committed:
Add track case for feature groups

iamgabrielma committed:
Add track case for loaded screens

iamgabrielma committed:
Add track cases for iap errors

iamgabrielma committed:
remove unnecessary string

iamgabrielma committed:
Inject Analytics component into Upgrades vm

iamgabrielma committed:
Reference analytics prop from UpgradesView

iamgabrielma committed:
Call planUpgradeScreenLoaded event on loaded view

iamgabrielma committed:
Hide analytics details from view

iamgabrielma committed:
Pass string to feature screen loaded details

iamgabrielma committed:
Track load and dismiss for UpgradeWaitingView

iamgabrielma committed:
whitespace

iamgabrielma committed:
Add track for plan details dismiss once view is loaded

iamgabrielma committed:
whitespace

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Upgrades/WooPlanFeatureBenefitsView.swift:
```diff
@@ -35,6 +35,10 @@ struct WooPlanFeatureBenefitsView: View {
         .navigationBarTitleDisplayMode(.inline)
+        .onAppear {
+            ServiceLocator.analytics.track(event:
+                    .InAppPurchases.planUpgradeFeatureScreenLoaded(featureGroup: wooPlanFeatureGroup.title))
```
I'm not entirely positive if this is the best way to pass the featureGroup parameter here. Initially I created an enum on `WooAnalyticsEvent.InAppPurchases` with the different cases, but filtering it depending on the view we're loading seems to be trickier than I initially expected. Happy to iterate further if you think we should pass other than the `wooPlanFeatureGroup.title`

joshheald commented on WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift:
```diff
@@ -67,6 +71,7 @@ final class UpgradesViewModel: ObservableObject {
 
         if let site = ServiceLocator.stores.sessionManager.defaultSite, !site.isSiteOwner {
             self.upgradeViewState = .prePurchaseError(.userNotAllowedToUpgrade)
+            analytics.track(event: .InAppPurchases.planUpgradePurchaseFailed(error: .userNotAllowedToUpgrade))
```
We could simplify this by having another subscriber to the `upgradeViewState` which logs the analytics as appropriate... that way we don't need all the repetition of the error type that we just set, we can just switch over the state and log the `purchaseFailed(error: error)` for all errors.

iamgabrielma replied:
Looks much better now, done on 5295571dae62443bc183e8bd8d1be50615141133

joshheald commented on WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift:
```diff
@@ -65,21 +65,30 @@ struct UpgradesView: View {
                             await upgradesViewModel.purchasePlan(with: plan.wpComPlan.id)
                         }
+                    }, onDismiss: {
+                        upgradesViewModel.trackDismiss(step: .planDetails)
```
This happens too often. I suspect it's on the wrong `onDismiss` modifier (because OwnerUpgradesView is a component of the screen which gets swapped out as the state changes, it's not the screen itself), or `onDismiss` doesn't work as we'd hope it would.

Ideally, we only want this to be tracked when the view actually slides off screen... currently it happens when the plan feature benefits are opened, when the waiting state starts, and possibly other times too.

joshheald commented on WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift:
```diff
@@ -457,6 +479,12 @@ struct UpgradeWaitingView: View {
 
             Spacer()
         }
+        .onAppear {
+            onViewLoaded()
+        }
+        .onDisappear {
+            onViewDismissed()
+        }
```
This looks like the cause of the over-logging of the events... `appear` and `disappear` will happen much more often than we want to track.

joshheald commented on WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift:
```diff
@@ -438,6 +458,8 @@ private extension PurchaseUpgradeError {
 
 struct UpgradeWaitingView: View {
     let planName: String
+    var onViewLoaded: (() -> Void)
+    var onViewDismissed: (() -> Void)
```
Why the naming difference here compared to `onDismiss` on the other views?

iamgabrielma replied:
No reason 😅

joshheald reviewed (commented):
I can see the events being logged.. but the loaded/dismissed events are being logged too frequently. I'm going to look for some better triggers that we may be able to use...

I also added a suggestion about neatening up the error tracking in the view model, by having the viewModel subscribe (for analytics) to the view state.

joshheald committed:
Replace deprecated presentationMode with dismissAction

joshheald committed:
10023 Track upgrade dismissal centrally

This isn’t the ideal way to track dismissal of these screens, but I can’t find a better one which takes account of swipe to dismiss, as long as we’re using a HostingController.

The ViewModel probably needs to provide a function for `didInteractivelyDismiss`, which is then made available by the HostingController, and called from UIKit when the ViewController/TransitionDelegate handles a swipe. But this does the trick, for the most part, and is a lot quicker.

joshheald committed:
10023 Move processingLoaded event to view model

joshheald committed:
Rename `waiting` step to match other `processing` events

joshheald commented:
> I can see the events being logged.. but the loaded/dismissed events are being logged too frequently. I'm going to look for some better triggers that we may be able to use...

From what I can tell, this is a pain in the neck to do while including the swipe to dismiss handler, because we're not using SwiftUI to do the modal presentation, we're using a hosting controller. `onAppear/onDisappear` do work for the most part, but they need to be on the top level view and will even then be called too frequently... just a lot less frequently.

50646ca has the dismissal approach, and 0e0765e has a move of the `.processingLoaded` to a subscription on the `viewState`, as suggested in my review – do you want to take a look at what else could be moved to that same switch statement?

Also, I think there's a weakness in our error logging – we don't log `purchaseUpgradeError`s (post-purchase) at all unless they're unrecognised, and if they're unrecognised, we don't log any detail of what the error was. We'll definitely need that for all post-purchase errors, and more detail about what they are. For recognised errors, we could log the errorCode, and a name where possible would be helpful for readability. It's also worth passing the underlying error through to analytics, we have an optional `error` parameter on `WooAnalyticsEvent.init` which will track some details using default formatting, that may be enough for unknown.

iamgabrielma committed:
Move error tracking to viewmodel state observer

iamgabrielma committed:
Pass generic Error to track event to filter necessary info

iamgabrielma committed:
Move .loaded case to model state observer

iamgabrielma committed:
Move .completed case to model state observer

iamgabrielma commented:
Thanks for working on this 🙇 

> [50646ca](https://github.com/woocommerce/woocommerce-ios/commit/50646cad08baf71d79dbe526fe005c27fb5b8702) has the dismissal approach, and [0e0765e](https://github.com/woocommerce/woocommerce-ios/commit/0e0765e0850696b6d721a44b1a6fc8f8ec6063d0) has a move of the `.processingLoaded` to a subscription on the `viewState`, as suggested in my review – do you want to take a look at what else could be moved to that same switch statement?

With the latest changes, along with the error tracking I moved the cases for `.loaded` and `.completed` as well to the switch.

> Also, I think there's a weakness in our error logging – we don't log `purchaseUpgradeError`s (post-purchase) at all unless they're unrecognised, and if they're unrecognised, we don't log any detail of what the error was. We'll definitely need that for all post-purchase errors, and more detail about what they are. For recognised errors, we could log the errorCode, and a name where possible would be helpful for readability. It's also worth passing the underlying error through to analytics, we have an optional `error` parameter on `WooAnalyticsEvent.init` which will track some details using default formatting, that may be enough for unknown.

With 5295571dae62443bc183e8bd8d1be50615141133 and c72eb2e5c6acfa048369c8d6328f2235c7e63b6b I separated the error tracking methods into pre and post-purchase, as we do in the UpgradesViewModel, and pass the generic error inside, filtering later on based on the error instance we pass through.

One issue I found is that the case for `inAppPurchaseFailed(WooWPComPlan, InAppPurchaseStore.Errors)` we would be passing the `WooWPComPlan` details as well, resulting in something like this being logged:

<img width="936" alt="Screenshot 2023-06-27 at 13 45 39" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/6afdf8bf-de5a-45de-b3cd-eb5c120357da">

I resolved it by filtering down what we actually pass as the error [here](https://github.com/woocommerce/woocommerce-ios/blob/c72eb2e5c6acfa048369c8d6328f2235c7e63b6b/WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift#L38C27-L47).

Here's the resulting message for different types of errors (pre-purchase, purchase known error, and purchase unknown error):

1. For `.prePurchaseError(.fetchError)`

```
2023-06-27 12:08:30.947197+0700 WooCommerce[11648:74404564] 🔵 Tracked plan_upgrade_purchase_failed, properties: [AnyHashable("error_code"): "0", AnyHashable("error_domain"): "WooCommerce.PrePurchaseError", AnyHashable("error"): "The operation couldn’t be completed. (WooCommerce.PrePurchaseError error 0.)", AnyHashable("error_description"): "WooCommerce.PrePurchaseError.fetchError", AnyHashable("blog_id"): 220618812, AnyHashable("is_wpcom_store"): true]
```

2. For `.purchaseUpgradeError(.inAppPurchaseFailed(plan, .transactionMissingAppAccountToken))`

```
2023-06-27 12:03:02.792131+0700 WooCommerce[11341:74399857] 🔵 Tracked plan_upgrade_purchase_failed, properties: [AnyHashable("error_domain"): "Yosemite.InAppPurchaseStore.Errors", AnyHashable("error"): "Purchase transaction missing account information", AnyHashable("is_wpcom_store"): true, AnyHashable("error_description"): "Yosemite.InAppPurchaseStore.Errors.transactionMissingAppAccountToken", AnyHashable("error_code"): "3", AnyHashable("blog_id"): 220618812]
```

3. For `.purchaseUpgradeError(.unknown)`

```
2023-06-27 12:10:31.795609+0700 WooCommerce[11738:74406845] 🔵 Tracked plan_upgrade_purchase_failed, properties: [AnyHashable("error"): "The operation couldn’t be completed. (WooCommerce.PurchaseUpgradeError error 2.)", AnyHashable("error_code"): "2", AnyHashable("error_description"): "WooCommerce.PurchaseUpgradeError.unknown", AnyHashable("blog_id"): 220618812, AnyHashable("error_domain"): "WooCommerce.PurchaseUpgradeError", AnyHashable("is_wpcom_store"): true]
```

joshheald reviewed (approved):
Nice one 👍
